Source/VCGeneration/CheckerPool.cs

```diff
@@ -10,69 +8,54 @@ namespace VC
   {
-    private readonly ConcurrentStack<Checker> availableCheckers = new();
-    private readonly ConcurrentQueue<TaskCompletionSource<Checker>> checkerWaiters = new();
-    private int notCreatedCheckers;
+    // Holds both created and not yet created checkers.
+    // Created checkers are kept at the front and uncreated ones at the end.
+    private readonly AsyncQueue<Checker?> checkerLine = new();
     private bool disposed;
 
     public VCGenOptions Options { get; }
 
     public CheckerPool(VCGenOptions options)
     {
-      this.Options = options;
-      notCreatedCheckers = options.VcsCores;
+      Options = options;
+      for (var index = 0; index < options.VcsCores; index++) {
+        checkerLine.Enqueue(null);
+      }
     }
 
-    public Task<Checker> FindCheckerFor(ConditionGeneration vcgen, Split split = null)
+    public async Task<Checker> FindCheckerFor(ConditionGeneration vcgen, Split? split, CancellationToken cancellationToken)
     {
       if (disposed) {
-        return Task.FromException<Checker>(new Exception("CheckerPool was already disposed"));
-      }
-        
-      if (availableCheckers.TryPop(out var result)) {
-        PrepareChecker(vcgen.program, split, result);
-        Contract.Assert(result != null);
-        return Task.FromResult(result);
+        throw new Exception("CheckerPool was already disposed");
       }
 
-      var afterDecrement = Interlocked.Decrement(ref notCreatedCheckers);
-      if (afterDecrement >= 0) {
-        var checker = CreateNewChecker();
-        PrepareChecker(vcgen.program, split, checker);
-        Contract.Assert(checker != null);
-        return Task.FromResult(checker);
-      }
-      Interlocked.Increment(ref notCreatedCheckers);
+      var checker = await checkerLine.Dequeue(cancellationToken) ?? CreateNewChecker();
 
-      var source = new TaskCompletionSource<Checker>();
-      checkerWaiters.Enqueue(source);
-      return source.Task.ContinueWith(task =>
-      {
-        PrepareChecker(vcgen.program, split, task.Result);
-        Contract.Assert(task.Result != null);
-        return task.Result;
-      });
+      PrepareChecker(vcgen.program, split, checker);
+      return checker;
     }
 
+    private int createdCheckers;
     private Checker CreateNewChecker()
     {
       var log = Options.ProverLogFilePath;
-      if (log != null && !log.Contains("@PROC@") && availableCheckers.Count > 0) {
-        log = log + "." + availableCheckers.Count;
+      var index = Interlocked.Increment(ref createdCheckers) - 1;
+      if (log != null && !log.Contains("@PROC@") && index > 0) {
+        log = log + "." + index;
       } 
 
       return new Checker(this, log, Options.ProverLogFileAppend);
     }
 
     public void Dispose()
     {
-      while (availableCheckers.TryPop(out var checker)) {
-        checker.Close();
-      }
       disposed = true;
+      foreach (var checker in checkerLine.ClearItems()) {
+        checker?.Close();
+      }
```

Comment:
Good point. I think in practice there's no issue because CheckerPool is only disposed when all work threads have already been awaited, so no concurrent calls to `CheckerPool.AddChecker` and `CheckerPool.Dispose` are done, but I'll add a lock anyways.

Reply:
👍